### PR TITLE
Physics V2 ragdolls

### PR DESCRIPTION
**Pull request by CedricGuillemet**
Experimental

image

Test PG : https://playground.babylonjs.com/#PS5DF6

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14606/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14606/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14606/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14606/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14606/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14606/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14606/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14606/merge#BCU1XR#0

File: packages/dev/core/src/Physics/v2/ragdoll.ts [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+                    {
+                        mass: mass,
+                        restitution: restitution,
+                        friction: 0.6,
+                    },
+                    this._scene
+                );
+                aggregate.body.setCollisionCallbackEnabled(true);
+                aggregate.body.disablePreStep = false;
+                (aggregate.body as any).collisionArray = new Array<number>(1);
```
Comment (RaananW) [inline packages/dev/core/src/Physics/v2/ragdoll.ts:187]: why the cast? is it not available on the class? Should it be added?

File: packages/dev/core/src/Physics/v2/ragdoll.ts [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+        if (newRotQuat) {
+            this._bones[boneIndex].setRotationQuaternion(newRotQuat, Space.WORLD, this._mesh);
+        }
+    }
+
+    // Return true if root bone is valid/exists in this.bonesNames. false otherwise.
+    private _defineRootBone(): boolean {
+        const skeletonRoots = this._skeleton.getChildren();
+        if (skeletonRoots.length != 1) {
+            console.log("Ragdoll creation failed: there can only be one root in the skeleton.");
```
Comment (RaananW) [inline packages/dev/core/src/Physics/v2/ragdoll.ts:272]: use Logger instead of console.log

File: packages/dev/core/src/Physics/v2/ragdoll.ts [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+            return false;
+        }
+
+        this._rootBoneName = skeletonRoots[0].name;
+        this._rootBoneIndex = this._boneNames.indexOf(this._rootBoneName);
+        if (this._rootBoneIndex == -1) {
+            console.log("Ragdoll creation failed: the array boneNames doesn't have the root bone. The root bone is", this._skeleton.getChildren());
```
Comment (RaananW) [inline packages/dev/core/src/Physics/v2/ragdoll.ts:279]: Same

File: packages/dev/core/src/Physics/v2/ragdoll.ts [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+            return false;
+        }
+
+        return true;
+    }
+
+    private _findNearestParent(boneIndex: number): any {
+        let nearestParent: Nullable<Bone> | undefined = this._bones[boneIndex].getParent();
+        do {
+            if (nearestParent != null && (this._boneNames as any).includes(nearestParent.name)) {
```
Comment (RaananW) [inline packages/dev/core/src/Physics/v2/ragdoll.ts:289]: are we missing the declaration for Array.prototype.includes?

File: packages/dev/core/src/Physics/v2/ragdoll.ts [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+                currentRagdollBoneProperties.joint = config[i].joint !== undefined ? config[i].joint : this._defaultJoint;
+                currentRagdollBoneProperties.rotationAxis = config[i].rotationAxis !== undefined ? config[i].rotationAxis : Axis.X;
+                currentRagdollBoneProperties.min = config[i].min !== undefined ? config[i].min : this._defaultJointMin;
+                currentRagdollBoneProperties.max = config[i].max !== undefined ? config[i].max : this._defaultJointMax;
+
+                // Offset value.
+                let boxOffset = 0;
+                if ((config[i].putBoxInBoneCenter !== undefined && config[i].putBoxInBoneCenter) || this._putBoxesInBoneCenter) {
+                    if (currentBone.length === undefined) {
+                        console.log("The length property is not defined for bone", currentBone.name);
```
Comment (carolhmj) [inline packages/dev/core/src/Physics/v2/ragdoll.ts:158]: We should use the Logger class and warn

Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/ragdoll.ts:158]: Fixed.

File: packages/dev/core/src/Physics/v2/ragdoll.ts (hunk 2 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/ragdoll.ts:272]: Fixed.

File: packages/dev/core/src/Physics/v2/ragdoll.ts (hunk 3 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/ragdoll.ts:279]: Fixed.

File: packages/dev/core/src/Physics/v2/ragdoll.ts (hunk 4 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/ragdoll.ts]: Fixed.

File: packages/dev/core/src/Physics/v2/ragdoll.ts (hunk 1 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/ragdoll.ts]: Fixed.

**Review by RaananW — APPROVED** [review]

**Pull request merged**